src/Microsoft.AspNet.Server.Kestrel/Http/SocketOutput.cs

```diff
@@ -91,6 +92,21 @@ public class SocketOutput : ISocketOutput
                 if (buffer.Count > 0)
                 {
                     var tail = ProducingStart();
+                    if (tail.IsDefault)
+                    {
+                        if (cancellationToken.CanBeCanceled)
+                        {
+                            return TaskUtilities.GetCancelledTask(
+                                cancellationToken.IsCancellationRequested ?
+                                cancellationToken :
+                                new CancellationToken(true));
```

Comment:
This is something I was really on the fence on too, but @Tratcher convinced me we should only cancel the task if the passed in token has actually been canceled.

The idea is that it's easier for the caller to reason about their code if the only way a write can fail is if the token that was passed in was actually canceled.

In theory, the behavior should be the same either way if the caller is always passing in the `RequestAborted` token. We just have to be extra careful to trip the token before determining whether to cancel or complete or cancel the task.

If you're not on board with this, we could probably wait until @Tratcher chimes in so we can have a group discussion. I'm probably not going to merge this until tomorrow anyway. Everyone here in Seattle has MLK day off :grin:

I think this PR is very close to being ready though.